### ISSUE TITLE
Clone example doesn't work

### ISSUE THREAD

**Issue by faxioman** [jelmer/dulwich#382]
Running the command:

```
python examples/clone.py git://github.com/jelmer/dulwich dulwich-clone
```

I get the error "TypeError: /jelmer/dulwich".
Python version: 3.4.3.
Platform: OSX Yosemite

**Comment by myint** [comment]
I get the same error under Python 3.4 and 3.5.

```
$ python examples/clone.py git@github.com:jelmer/dulwich.git ssh_clone_test8
Traceback (most recent call last):
  File "examples/clone.py", line 18, in <module>
    porcelain.clone(args[1], args[2])
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/porcelain.py", line 253, in clone
    progress=errstream.write)
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 242, in fetch
    progress)
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 546, in fetch_pack
    proto, can_read = self._connect(b'upload-pack', path)
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 909, in _connect
    raise TypeError(path)
TypeError: jelmer/dulwich.git
```

**Comment by jelmer** [comment]
On Sat, Oct 24, 2015 at 09:40:00AM -0700, Steven Myint wrote:
> I get the same error under Python 3.4 and 3.5.
> 
> ```
> $ python examples/clone.py git@github.com:jelmer/dulwich.git ssh_clone_test8
> Traceback (most recent call last):
>   File "examples/clone.py", line 18, in <module>
>     porcelain.clone(args[1], args[2])
>   File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/porcelain.py", line 253, in clone
>     progress=errstream.write)
>   File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 242, in fetch
>     progress)
>   File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 546, in fetch_pack
>     proto, can_read = self._connect(b'upload-pack', path)
>   File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 909, in _connect
>     raise TypeError(path)
> TypeError: jelmer/dulwich.git

It's a Python-specific bug. We lack tests that exercise this specific code path, which is why it didn't come up during the Python3 porting effort.

Jelmer

**Comment by myint** [comment]
Makes sense.

I see different errors when cloning a local directory.

```
$ python examples/clone.py . /tmp/sdfsdfdsfsdfdd
Traceback (most recent call last):
  File "examples/clone.py", line 18, in <module>
    porcelain.clone(args[1], args[2])
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/porcelain.py", line 253, in clone
    progress=errstream.write)
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/client.py", line 792, in fetch
    progress=progress)
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/repo.py", line 231, in fetch
    progress))
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/object_store.py", line 386, in add_objects
    if len(objects) == 0:
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/object_store.py", line 885, in __len__
    return len(list(self.itershas()))
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/object_store.py", line 859, in itershas
    for sha in self.sha_iter:
  File "/private/tmp/dulwich/3.4/lib/python3.4/site-packages/dulwich-0.11.3-py3.4.egg/dulwich/object_store.py", line 1045, in next
    self.progress(message.encode('ascii'))
TypeError: must be str, not bytes
```

Is `progress` meant to be for `bytes` output or `str` output? The [default argument](https://github.com/jelmer/dulwich/blob/master/dulwich/porcelain.py#L218) for `progress` is `sys.stdout`, which wants `str`. But the unit test uses `BytesIO` for `progress`, which wants `bytes`.

**Comment by jelmer** [comment]
errstream is meant to be a bytestream (as we don't know the encoding the remote host is using).

**Comment by meatballs** [comment]
My first attempt to use Dulwich and I've hit this problem which appears to stem from the _connect methods in client.py having:
```
if type(path) is not bytes:
            raise TypeError(path)
```
I've tried passing the url as a bytestring, but that just gives the following error:
```
line 1151, in get_transport_and_path
    if ':' in location and not '@' in location:
TypeError: a bytes-like object is required, not 'str'
```

**Comment by jelmer** [comment]
On Sun, Jan 03, 2016 at 11:56:48AM -0800, Owen Campbell wrote:
> My first attempt to use Dulwich and I've hit this problem which appears to stem from the _connect methods in client.py having:
> ```
> if type(path) is not bytes:
>             raise TypeError(path)
> ```
> I've tried passing the url as a bytestring, but that just gives the following error:
> ```
> line 1151, in get_transport_and_path
>     if ':' in location and not '@' in location:
> TypeError: a bytes-like object is required, not 'str'
> ```

Unfortunately this example doesn't yet work with Python 3. There is an open issue about the conversion
of 'path' from unicode to bytestrings.

Jelmer

**Comment by stuaxo** [comment]
I've had another error trying to use clone in python 3

```
>>> porcelain.clone("git@bitbucket.org:infectious/ietl.git", "/tmp/git-cache/ietl.git")
```
```                                        
Traceback (most recent call last):
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py", line 249, in fetch
    progress)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py", line 555, in fetch_pack
    proto, can_read = self._connect(b'upload-pack', path)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py", line 914, in _connect
    raise TypeError(path)
TypeError: infectious/ietl.git

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/stu/.virtualenvs/juggler-ng/bin/jng", line 9, in <module>
    load_entry_point('juggler-ng', 'console_scripts', 'jng')()
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/stu/projects/work/infectiousmedia/ng/juggler_ng/cmdline/__init__.py", line 127, in run_all
    job.run()
  File "/home/stu/projects/work/infectiousmedia/ng/juggler_ng/jobs/scheduled_job.py", line 115, in run
    return self.f(*self.fargs, **self.fkwargs)
  File "/home/stu/projects/work/infectiousmedia/ng/juggler_ng/jobs/loader/ietl_job.py", line 200, in run_module_job
    setup_env()
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/kids/cache/__init__.py", line 112, in _cache_wrapper
    result = wrapped(*args, **kwargs)
  File "/home/stu/projects/work/infectiousmedia/ng/juggler_ng/jobs/loader/ietl_job.py", line 171, in setup_env
    ietl_dir, updated = git.get_project(settings.jobs.ietl_repo, pull=True, cache_dir=git_root)
  File "/home/stu/projects/work/infectiousmedia/ng/juggler_ng/git.py", line 88, in get_project
    porcelain.clone(uri, project_dir)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/porcelain.py", line 258, in clone
    progress=errstream.write)
  File "/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py", line 251, in fetch
    abort()
UnboundLocalError: local variable 'abort' referenced before assignment

```

**Comment by jelmer** [comment]
What version of Dulwich is this? There was a fix related to abort recently.

On 5 April 2016 14:24:36 CEST, Stuart Axon <notifications@github.com> wrote:
>I've had another error trying to use clone in python 3
>
>```
>>>> porcelain.clone("git@bitbucket.org:infectious/ietl.git",
>"/tmp/git-cache/ietl.git")
>```
>```                                        
>Traceback (most recent call last):
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py",
>line 249, in fetch
>    progress)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py",
>line 555, in fetch_pack
>    proto, can_read = self._connect(b'upload-pack', path)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py",
>line 914, in _connect
>    raise TypeError(path)
>TypeError: infectious/ietl.git
>
>During handling of the above exception, another exception occurred:
>
>Traceback (most recent call last):
>  File "/home/stu/.virtualenvs/juggler-ng/bin/jng", line 9, in <module>
>    load_entry_point('juggler-ng', 'console_scripts', 'jng')()
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py",
>line 716, in __call__
>    return self.main(*args, **kwargs)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py",
>line 696, in main
>    rv = self.invoke(ctx)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py",
>line 1060, in invoke
>    return _process_result(sub_ctx.command.invoke(sub_ctx))
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py",
>line 889, in invoke
>    return ctx.invoke(self.callback, **ctx.params)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/click/core.py",
>line 534, in invoke
>    return callback(*args, **kwargs)
>File
>"/home/stu/projects/work/infectiousmedia/ng/juggler_ng/cmdline/__init__.py",
>line 127, in run_all
>    job.run()
>File
>"/home/stu/projects/work/infectiousmedia/ng/juggler_ng/jobs/scheduled_job.py",
>line 115, in run
>    return self.f(*self.fargs, **self.fkwargs)
>File
>"/home/stu/projects/work/infectiousmedia/ng/juggler_ng/jobs/loader/ietl_job.py",
>line 200, in run_module_job
>    setup_env()
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/kids/cache/__init__.py",
>line 112, in _cache_wrapper
>    result = wrapped(*args, **kwargs)
>File
>"/home/stu/projects/work/infectiousmedia/ng/juggler_ng/jobs/loader/ietl_job.py",
>line 171, in setup_env
>ietl_dir, updated = git.get_project(settings.jobs.ietl_repo, pull=True,
>cache_dir=git_root)
>File "/home/stu/projects/work/infectiousmedia/ng/juggler_ng/git.py",
>line 88, in get_project
>    porcelain.clone(uri, project_dir)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/porcelain.py",
>line 258, in clone
>    progress=errstream.write)
>File
>"/home/stu/.virtualenvs/juggler-ng/lib/python3.5/site-packages/dulwich/client.py",
>line 251, in fetch
>    abort()
>UnboundLocalError: local variable 'abort' referenced before assignment
>
>```
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/jelmer/dulwich/issues/382#issuecomment-205777846

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by stuaxo** [comment]
Version 0.12.0

**Comment by jelmer** [comment]
At least the second exception is fixed in HEAD/0.13.0

On 5 April 2016 14:27:39 CEST, Stuart Axon <notifications@github.com> wrote:
>Version 0.12.0
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/jelmer/dulwich/issues/382#issuecomment-205778900

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by jelmer** [comment]
This was fixed sometime before 0.16.3.
